### PR TITLE
Disallow phase inconsistent inline parameters

### PR DESCRIPTION
**Pull request by nicolasstucki**
In preparation for #8060.

Originally we used `inline` parameters as the way to recover statically known values from the arguments. This worked on some predefined types such as literal primitives, case objects, case classes, enums among others. The latter ones where ad-hoc addons without a clear specification.

`ValueOfExpr` also recovers statically known values but it is more general. It can be implemented for any arbitrary type that represents a value. In this PR we introduce some extra ways to use this such as `Expr.value`, `matching.Value` and `matching.ValueSeq`.

`ValueOfExpr` makes the current use of inline parameters completely redundant hence this PR replaces all uses of inline parameters by equivalent code using `ValueOfExpr`.

It also simplifies the PCP as it removes a special case from the rules.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Will need to be rebased on #8077 (the second commit will disapear)

File: docs/docs/reference/metaprogramming/macros.md [hunk 1]
```diff
@@ -414,33 +414,26 @@ assume that both definitions are local.
...
-aspect is also important for macro expansion.  To illustrate this,
-consider an implementation of the `power` function that makes use of a
-statically known exponent:
+aspect is also important for macro expansion.
+
+To get values out of expressions containing constants `Expr` provides the methods
+`value` (or `getValue`). This will convert the `Expr[T]` into a `T` (or `Some[T]`) when the
+expression contains value. Otherwise it will emit an error (or retrun `None`).
+To avoid having incidental val bindings generated by the inlining of the `def`
+it is recomended to use an inline parameter. To illustrate this, consider an
```
Comment (liufengyun) [inline docs/docs/reference/metaprogramming/macros.md]: typo: `recomended ` -> `recommended `

File: docs/docs/reference/metaprogramming/macros.md [hunk 2]
```diff
@@ -414,33 +414,26 @@ assume that both definitions are local.
...
+implementation of the `power` function that makes use of a statically known exponent:
 ```scala
-inline def power(inline n: Int, x: Double) = ${ powerCode(n, 'x) }
+inline def power(x: Double, inline n: Int) = ${ powerCode('x, 'n) }
+
+private def powerCode(x: Expr[Double], n: Expr[Int]): Expr[Double] =
+  powerCode(x, n.value) // Transform `n` into an Int or emit an error if it is not possible
```
Comment (liufengyun) [inline docs/docs/reference/metaprogramming/macros.md]: Maybe do it safely by `n.getValue`

File: library/src/scala/quoted/ValueOfExpr.scala [hunk 3]
```diff
@@ -30,4 +32,252 @@ object ValueOfExpr {
...
 
+  given Option_delegate[T](given Type[T], ValueOfExpr[T]): ValueOfExpr[Option[T]] = new {
+    def apply(x: Expr[Option[T]])(given QuoteContext): Option[Option[T]] = x match {
+      case '{ None: Option[T] } => Some(None) // FIXME: remove ascription, Matcher should be able to match this expression
+      case '{ new Some[T](${Value(y)}) } => Some(Some(y))
+      case '{     Some[T](${Value(y)}) } => Some(Some(y))
+      case _ => None
+    }
+    override def toString(): String = "scala.quoted.ValueOfExpr.Option_delegate"
+  }
```
Comment (liufengyun) [inline library/src/scala/quoted/ValueOfExpr.scala]: Need to rename `*_delegate` , maybe in another PR.

File: library/src/scala/quoted/matching/Value.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+   *  ```
+   *  (x: Expr[B]) match {
+   *    case Value(value) => ... // value: B
+   *  }
+   *  ```
+   */
+  def unapply[T](expr: Expr[T])(given valueOf: ValueOfExpr[T], qxtc: QuoteContext): Option[T] =
+    valueOf(expr)
+
+}
```
Comment (liufengyun) [inline library/src/scala/quoted/matching/Value.scala]: We already have extractor for Const, a duplicate? The name `Value` is a little misleading, if we in fact only want to get the literal const out.

File: library/src/scala/quoted/Expr.scala [hunk 5]
```diff
@@ -18,6 +18,14 @@ class Expr[+T] private[scala] {
    */
   final def getValue[U >: T](given qctx: QuoteContext, valueOf: ValueOfExpr[U]): Option[U] = valueOf(this)
 
+  /** Return the value of this expression.
+   *
+   *  Emits an error error and throws if the expression does not contain a value or contains side effects.
+   *  Otherwise returns the value.
+   */
+  final def value[U >: T](given qctx: QuoteContext, valueOf: ValueOfExpr[U]): U =
+    valueOf(this).getOrElse(qctx.throwError(s"Expected a known value. \n\nThe value of: $show\ncould not be recovered using $valueOf", this))
```
Comment (liufengyun) [inline library/src/scala/quoted/Expr.scala]: It seems this method only invites errors in meta-programming, maybe we should just remove it.

File: library/src/scala/quoted/Expr.scala [hunk 6]
```diff
@@ -18,6 +18,14 @@ class Expr[+T] private[scala] {
    */
   final def getValue[U >: T](given qctx: QuoteContext, valueOf: ValueOfExpr[U]): Option[U] = valueOf(this)
```
Comment (liufengyun) [inline library/src/scala/quoted/Expr.scala]: I find `getValue` is a little misleading, if what it does is to get the literal out. Maybe `getLiteralValue`?

File: docs/docs/reference/metaprogramming/macros.md (hunk 2 above)
Comment (nicolasstucki) [reply docs/docs/reference/metaprogramming/macros.md]: This has the same semantics as before. If the inline parameter is not a constant, an error is emitted.

File: library/src/scala/quoted/ValueOfExpr.scala (hunk 3 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/ValueOfExpr.scala]: Agreed

File: library/src/scala/quoted/Expr.scala (hunk 6 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Expr.scala]: It is not necessarily a literal.

File: library/src/scala/quoted/matching/Value.scala (hunk 4 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/matching/Value.scala]: Yes, Value is more general. I plan to replace and remove Const in another PR

File: library/src/scala/quoted/Expr.scala (hunk 5 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Expr.scala]: Without it we would end up writing `expr.getValue.get` all over the place which would crash the macro expansion. This alternative emits an error saying that the call expected a constant value. Pointing to the argument that is not a value.

**Pull request merged**